### DISCUSSION TITLE
Slide joint actuator that lifts a hinged body

### DISCUSSION THREAD

**Discussion by lucasw** [google-deepmind/mujoco#390] [Asking for Help]
I'd like to make a linear actuator on a hinge that lifts a body that's constrained by a hinge.  I saw the slider crank example and tried that out but don't want to have the crank part of it and it is very jittery and moves the wrong way and breaks.  I next tried out using the slide joint and an equality weld to attach the end of the slide to the body- but for that the weld seems very soft, maybe I only need help with solver parameters to tighten it?

https://user-images.githubusercontent.com/1334122/179545918-47990f7f-8c7c-4857-8123-91f4195f7c96.mp4

```
<mujoco model="slide">
  <visual>
    <scale slidercrank=".1" />
    <rgba slidercrank="0.5 0.4 0.8 1" crankbroken="0.0 0.6 0.2 1" />
  </visual>

  <default>
    <position ctrllimited="true" ctrlrange="0.02 0.15" kp="10" />
    <default class="slidersite">
      <site type="cylinder" size=".01 .015" rgba="1 0 0 1" />
    </default>
    <default class="cranksite">
      <site size=".012" rgba="0 0 1 1" />
    </default>
  </default>

  <worldbody>
    <light pos="0 0 1"/>
    <geom type="plane" size=".5 .5 .01" pos="0 0 -.05" />

    <body pos="0.1 0 -0.025">
      <joint damping=".1" axis="0 1 0" />
      <geom type="cylinder" size=".006" fromto="0 0 0 0 0 .01" rgba=".9 .6 1 1" />
      <site name="slidersite0" class="slidersite" pos="0 0 0" zaxis="0 0 1" />
    </body>
    <body pos="0 0 0">
      <joint damping=".1" axis="0 1 0" />
      <geom type="capsule" size=".01" fromto="0 0 0 .15 0 0" />
      <site name="cranksite0" class="cranksite" pos="0.1 0 0" size=".012" />
    </body>

    <body pos="0.1 0.15 -0.04">
      <joint damping=".1" axis="0 1 0" />
      <geom type="cylinder" size=".006" fromto="0 0 0 0 0 .01" rgba=".9 .6 1 1" />
      <body name="slider_end0" pos="0 0 0">
        <joint name="slider_base" type="slide" damping=".1" axis="0 0 1" limited="true" range="0.02 0.5" />
        <geom type="cylinder" size=".006" fromto="0 0 0 0 0 .01" rgba=".9 .6 1 1" />
      </body>
    </body>
    <body pos="0 0.15 0">
      <joint damping=".1" axis="0 1 0" limited="true" range="-60 60" />
      <geom type="capsule" size=".01" fromto="0 0 0 .15 0 0" />
      <body name="slider_end1" pos="0.1 0.0 -0.015">
        <joint name="slider_end1" damping=".1" axis="0 1 0" />
        <geom type="cylinder" size=".007" fromto="0 0 0 0 0 -.008" rgba=".9 .3 .5 1" />
      </body>
    </body>

  </worldbody>

  <equality>
    <weld name="slider_end_connection" body1="slider_end0" body2="slider_end1" />
  </equality>

  <actuator>
    <position name="slidercrank0" cranksite="cranksite0" slidersite="slidersite0" cranklength=".02" />
    <position name="linear_actuator1" joint="slider_base" />
  </actuator>
</mujoco>
```
             

Similar to https://github.com/deepmind/mujoco/discussions/374#discussioncomment-3105459

**Comment by yuvaltassa** [comment]
I'm a bit confused by what you're trying to do, "a linear actuator on a hinge" does not compile 🙂 

That said, if you want to stiffen the [weld](https://mujoco.readthedocs.io/en/latest/XMLreference.html#equality-weld), use `solref` and `solimp` as usual, maybe something like `<weld ... solimp="0.9 .99 .0001" solref=".001 1"/>`?

**Reply by lucasw** [reply]
The weld with those solimp/solref parameters seems to work ok and I'll use it for now- but the bigger question is if the slide joint and weld is the best way to accomplish this or there is another joint or way to attach things together and actuate them roughly like a hydraulic cylinder.  Does the video or xml show what I'm trying to do better if I'm not describing it correctly?

I also tried out the muscle + tendon and it works well lifting the arm, but only seems to fall back down under gravity, I can't generate the opposite force (maybe I can attach another muscle and tendon above the arm to get the opposite force, but it is more complicated to control and resembles the hydraulic cylinder even less).

```
   ...
    <body pos="0.1 0.3 -0.04">
      <joint damping=".1" axis="0 1 0" />
      <geom type="cylinder" size=".006" fromto="0 0 0 0 0 .01" rgba=".9 .6 1 1" />
      <site name="tendon0" pos="0 0 0" size="0.013" />
    </body>
    <body pos="0 0.3 0" euler="0 -60 0">
      <joint name="tendon_arm_hinge" damping=".1" axis="0 1 0" limited="true" range="-90 90" />
      <geom type="capsule" size=".01" fromto="0 0 0 .15 0 0" />
      <site name="tendon1" pos="0.1 0 0" size="0.013" />
    </body>

  </worldbody>

  <tendon>
    <spatial name="arm_tendon" width="0.01" stiffness="0.1">
      <site site="tendon0" />
      <site site="tendon1" />
    </spatial>
  </tendon>

  <actuator>
   ...
    <muscle name="muscle" tendon="arm_tendon" ctrllimited="true" ctrlrange="-1.0 1.0" range="-1.0 1.0" />
  </actuator>
```

**Comment by KingBright** [comment]
Recently I just have met the same problem like yours, and for now I haven't found a better way to do that.

**Comment by yuvaltassa** [comment]
There was a latent bug in slider cranks that was recently fixed ([item 14](https://mujoco.readthedocs.io/en/latest/changelog.html#version-3-2-5-nov-4-2024) in the 3.2.5 changelog), perhaps take another look at these? Another possibility is to put a spatial tendon and actuate that. Hope this helps.
